src/compiler/checker.ts

```diff
@@ -32791,8 +32810,38 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function checkNonNullAssertion(node: NonNullExpression) {
-        return node.flags & NodeFlags.OptionalChain ? checkNonNullChain(node as NonNullChain) :
-            getNonNullableType(checkExpression(node.expression));
+        const exprType = checkExpression(node.expression);
+        const resultType = node.flags & NodeFlags.OptionalChain ? checkNonNullChain(node as NonNullChain) : getNonNullableType(exprType);
+        if (!isErrorType(resultType)) {
+            addLazyDiagnostic(() => {
+                if (isTypeIdenticalTo(exprType, resultType)) {

```

Comment:
Why can't you just check if any of the constituent types in `exprType` contain `TypeFlags.Null | TypeFlags.Undefined`?

Reply:
`T extends X | null | undefined`?

Reply:
Ugh - okay, I trust we have a test on that.